File: haskell/src/InfoVis/Parallel/Rendering/Buffers.hs
{-# LANGUAGE NamedFieldPuns  #-}
{-# LANGUAGE Rank2Types      #-}
{-# LANGUAGE RecordWildCards #-}


module InfoVis.Parallel.Rendering.Buffers (
  ShapeBuffer
, createShapeBuffer
, prepareShapeBuffer
, destroyShapeBuffer
, hasIdentifier
, insertPositions
, updatePositions
, updateRotations
, updateScales
, updateColor
, deleteInstance
, drawInstances
) where


import Control.Lens.Lens (Lens', lens)
import Control.Lens.Setter (over)
import Control.Monad (unless, when)
import Data.Array.Storable (getElems, newArray, newListArray, touchStorableArray, withStorableArray)
import Foreign.Ptr (plusPtr)
import Foreign.Storable (Storable, poke, sizeOf)
import Graphics.GL.Types (GLfloat, GLuint)
import Graphics.Rendering.OpenGL.GL (($=!), deleteObjectName, deleteObjectNames, genObjectName)
import Graphics.Rendering.OpenGL.GL.BufferObjects (BufferAccess(WriteOnly), BufferObject, BufferTarget(..), BufferUsage(DynamicDraw), TransferDirection(ReadFromBuffer), bindBuffer, bufferData, bufferSubData, withMappedBuffer)
import Graphics.Rendering.OpenGL.GL.PrimitiveMode (PrimitiveMode(..))
import Graphics.Rendering.OpenGL.GL.Tensor (Vector3(..), Vector4(..), Vertex3(..))
import Graphics.Rendering.OpenGL.GL.VertexArrays (NumArrayIndices, NumInstances, drawArraysInstanced)
import InfoVis.Parallel.NewTypes (Identifier)
import InfoVis.Parallel.Rendering.Program (ShapeProgram, bindColors, bindMesh, bindPositions, bindRotations, bindScales, selectShapeProgram, setProjectionModelView)
import Linear.Matrix (M44)
import System.IO (hPrint, stderr)

import qualified Data.IntMap.Lazy as IM (IntMap, empty, findMin, fromList, insert, null, toList, union)
import qualified Data.Map.Strict as M (Map, delete, empty, findWithDefault, foldl, insertWith, member)
import qualified Data.IntSet as IS (IntSet, deleteFindMin, empty, findMax, foldl, fromList, null, singleton, toList, union)


type Position = Vertex3 GLfloat


zeroPosition :: Position
zeroPosition = Vertex3 0 0 0


type Rotation = Vector4 GLfloat


zeroRotation :: Rotation
zeroRotation = Vector4 0 0 0 1


type Scale = Vector3 GLfloat


zeroScale :: Scale
zeroScale = Vector3 0 0 0


type Color = GLuint


zeroColor :: Color
zeroColor = 0


data ShapeBuffer =
  ShapeBuffer
  {
    shapeProgram     :: ShapeProgram
  , primitiveMode    :: PrimitiveMode
  , mesh             :: Either (IO BufferObject) BufferObject
  , vertexCount      :: NumArrayIndices
  , instanceCount    :: NumInstances
  , positions        :: BufferObject
  , rotations        :: BufferObject
  , scales           :: BufferObject
  , colors           :: BufferObject
  , size             :: Int
  , empties          :: IS.IntSet
  , locationses      :: M.Map Identifier IS.IntSet
  , pendingPositions :: IM.IntMap Position
  , pendingRotations :: IM.IntMap Rotation
  , pendingScales    :: IM.IntMap Scale
  , pendingColors    :: IM.IntMap Color
  , pendingSize      :: Int
  }


positionsLens :: Lens' ShapeBuffer (IM.IntMap Position)
positionsLens = lens pendingPositions (\s pendingPositions -> s {pendingPositions})


rotationsLens :: Lens' ShapeBuffer (IM.IntMap Rotation)
rotationsLens = lens pendingRotations (\s pendingRotations -> s {pendingRotations})


scalesLens :: Lens' ShapeBuffer (IM.IntMap Scale)
scalesLens = lens pendingScales (\s pendingScales -> s {pendingScales})


colorsLens :: Lens' ShapeBuffer (IM.IntMap Color)
colorsLens = lens pendingColors (\s pendingColors -> s {pendingColors})


hasIdentifier :: ShapeBuffer
              -> Identifier
              -> Bool
hasIdentifier = flip M.member . locationses


createShapeBuffer :: ShapeProgram
                  -> (PrimitiveMode, [Position])
                  -> ShapeBuffer
createShapeBuffer shapeProgram (primitiveMode, primitives) =
  let
    mesh      = Left $ buildBuffer primitives
    positions = undefined
    rotations = undefined
    scales    = undefined
    colors    = undefined
    vertexCount   = fromIntegral $ length primitives
    instanceCount = 0
    size = 0
    empties = IS.empty
    locationses = M.empty
    pendingPositions = IM.empty
    pendingRotations = IM.empty
    pendingScales    = IM.empty
    pendingColors    = IM.empty
    pendingSize = size
  in
    ShapeBuffer{..}


destroyShapeBuffer :: ShapeBuffer
                   -> IO ()
destroyShapeBuffer ShapeBuffer{..}
  | size == 0 = return ()
  | otherwise = case mesh of
                  Left  _     -> deleteObjectNames [       positions, rotations, scales, colors]
                  Right mesh' -> deleteObjectNames [mesh', positions, rotations, scales, colors]


insertPositions :: (Identifier, [Position])
                -> ShapeBuffer
                -> ShapeBuffer
insertPositions = flip $ (. uncurry ((<$>) . (,))) . foldl insertPosition


insertPosition :: ShapeBuffer
               -> (Identifier, Position)
               -> ShapeBuffer
insertPosition shapeBuffer@ShapeBuffer{..} (identifier, vertex) =
  let
    (empties', pendingSize'') =
      if IS.null empties
        then let
               pendingSize' = 3 * pendingSize `div` 2 + 1
             in
               (
                 IS.fromList [pendingSize..(pendingSize'-1)]
               , pendingSize'
               )
        else (empties, pendingSize)
    (location, empties'') = IS.deleteFindMin empties'
  in
    shapeBuffer
    {
      empties          = empties''
    , locationses      = M.insertWith IS.union identifier (IS.singleton location) locationses
    , pendingPositions = IM.insert location vertex pendingPositions
    , pendingSize      = pendingSize''
    }


updatePositions :: (Identifier, [Position])
                -> ShapeBuffer
                -> ShapeBuffer
updatePositions = updateAttributes positionsLens

  
updateRotations :: (Identifier, [Rotation])
                -> ShapeBuffer
                -> ShapeBuffer
updateRotations = updateAttributes rotationsLens
 

updateScales :: (Identifier, [Scale])
             -> ShapeBuffer
             -> ShapeBuffer
updateScales = updateAttributes scalesLens
 

updateAttributes :: Lens' ShapeBuffer (IM.IntMap a)
                 -> (Identifier, [a])
                 -> ShapeBuffer
                 -> ShapeBuffer
updateAttributes field (identifier, values) shapeBuffer@ShapeBuffer{..} =
  over
    field
    (
      IM.union
        $ IM.fromList (zip (IS.toList $ M.findWithDefault IS.empty identifier locationses) values)
    )
    shapeBuffer


updateColor :: (Identifier, Color)
            -> ShapeBuffer
            -> ShapeBuffer
updateColor = updateAttribute colorsLens
 

updateAttribute :: Lens' ShapeBuffer (IM.IntMap a)
                -> (Identifier, a)
                -> ShapeBuffer
                -> ShapeBuffer
updateAttribute field (identifier, value) shapeBuffer@ShapeBuffer{..} =
  over
    field
    (
      flip
        (IS.foldl (flip $ flip IM.insert value))
        $ M.findWithDefault IS.empty identifier locationses
    )
    shapeBuffer


deleteInstance :: ShapeBuffer
               -> Identifier
               -> ShapeBuffer
deleteInstance shapeBuffer@ShapeBuffer{..} identifier =
  let
    locations = M.findWithDefault IS.empty identifier locationses
  in
    shapeBuffer
    {
      empties       = empties `IS.union` locations
    , locationses   = M.delete identifier locationses
    , pendingScales = IS.foldl (flip $ flip IM.insert zeroScale) pendingScales locations
    }


prepareShapeBuffer :: ShapeBuffer
                   -> IO ShapeBuffer
prepareShapeBuffer = (>>= updateShapeBuffer) . expandShapeBuffer


updateShapeBuffer :: ShapeBuffer
                  -> IO ShapeBuffer
updateShapeBuffer shapeBuffer@ShapeBuffer{..} =
  if IM.null pendingPositions && IM.null pendingRotations && IM.null pendingScales && IM.null pendingColors
    then return shapeBuffer
    else do
           pendingPositions `updateBuffer` positions
           pendingRotations `updateBuffer` rotations
           pendingScales    `updateBuffer` scales
           pendingColors    `updateBuffer` colors
           return
             shapeBuffer
             {
               instanceCount    = fromIntegral $ IS.findMax (M.foldl IS.union (IS.singleton (-1)) locationses) + 1
             , pendingPositions = IM.empty
             , pendingRotations = IM.empty
             , pendingScales    = IM.empty
             , pendingColors    = IM.empty
             }


expandShapeBuffer :: ShapeBuffer
                  -> IO ShapeBuffer
expandShapeBuffer shapeBuffer@ShapeBuffer{..} =
  if pendingSize <= size
    then return shapeBuffer
    else do
           mesh' <- either id return mesh
           positions' <- if size == 0 then buildBuffer $ replicate pendingSize zeroPosition else expandBuffer zeroPosition size pendingSize positions
           rotations' <- if size == 0 then buildBuffer $ replicate pendingSize zeroRotation else expandBuffer zeroRotation size pendingSize rotations
           scales'    <- if size == 0 then buildBuffer $ replicate pendingSize zeroScale    else expandBuffer zeroScale    size pendingSize scales
           colors'    <- if size == 0 then buildBuffer $ replicate pendingSize zeroColor    else expandBuffer zeroColor    size pendingSize colors
           return 
             shapeBuffer
             {
               mesh      = Right mesh'
             , size      = pendingSize
             , positions = positions'
             , rotations = rotations'
             , scales    = scales'
             , colors    = colors'
             } 


drawInstances :: M44 GLfloat
              -> M44 GLfloat
              -> ShapeBuffer
              -> IO ()
drawInstances projection modelView ShapeBuffer{..} =
  when (size > 0)
    $ do
    let
      Right mesh' = mesh
    selectShapeProgram $ Just shapeProgram
    setProjectionModelView shapeProgram projection modelView 
    shapeProgram `bindMesh`      Just mesh'
    shapeProgram `bindPositions` Just positions
    shapeProgram `bindRotations` Just rotations
    shapeProgram `bindScales`    Just scales
    shapeProgram `bindColors`    Just colors
    drawArraysInstanced primitiveMode 0 vertexCount instanceCount
    selectShapeProgram Nothing


buildBuffer :: Storable a
            => [a]
            -> IO BufferObject
buildBuffer primitives =
  do
    let
      stride = sizeOf $ head primitives
      size = length primitives
    bufferObject <- genObjectName
    bindBuffer ArrayBuffer $=! Just bufferObject
    values <- newListArray (0, size) primitives
    withStorableArray values
      $ \ptr ->
        bufferData ArrayBuffer $=! (fromIntegral $ size * stride, ptr, DynamicDraw)
    bindBuffer ArrayBuffer $=! Nothing
    return bufferObject


updateBuffer :: Storable a
             => IM.IntMap a
             -> BufferObject
             -> IO ()
updateBuffer updates bufferObject =
  unless (IM.null updates)
    $ do
      let
        stride = sizeOf . snd $ IM.findMin updates
      bindBuffer ArrayBuffer $=! Just bufferObject
      withMappedBuffer ArrayBuffer WriteOnly
        (
          \ptr ->
            sequence_
              [
                ptr `plusPtr` (location * stride) `poke` value
              |
                (location, value) <- IM.toList updates
              ]
        )
        $ hPrint stderr
      bindBuffer ArrayBuffer $=! Nothing


expandBuffer :: Storable a
             => a
             -> Int
             -> Int
             -> BufferObject
             -> IO BufferObject
expandBuffer exemplar oldSize newSize oldBufferObject =
  do
    let
      stride = sizeOf exemplar
    bindBuffer ArrayBuffer $=! Just oldBufferObject
    oldValues <- newArray (0, oldSize) exemplar
    touchStorableArray oldValues
    withStorableArray oldValues
      $ bufferSubData ArrayBuffer ReadFromBuffer 0 (fromIntegral $ oldSize * stride)
    newValues <- getElems oldValues >>= newListArray (0, newSize) . (++ repeat exemplar)
    deleteObjectName oldBufferObject
    newBufferObject <- genObjectName
    bindBuffer ArrayBuffer $=! Just newBufferObject
    withStorableArray newValues
      $ \ptr ->
        bufferData ArrayBuffer $=! (fromIntegral $ newSize * stride, ptr, DynamicDraw)
    bindBuffer ArrayBuffer $=! Nothing
    return newBufferObject

File: web/src/rendering/buffers.js

const Program = require("./program")

require("../gl-matrix")


const zeroPosition = glMatrix.vec3.fromValues(0, 0, 0)

const zeroRotation = glMatrix.quat.fromValues(0, 0, 0, 1)

const zeroScale = glMatrix.vec3.fromValues(0, 0, 0)

const zeroColor = 0.0


const positionsLens = {
  get : (shapeBuffer                  ) => shapeBuffer.pendingPositions
, set : (shapeBuffer, pendingPositions) => shapeBuffer.pendingPositions = pendingPositions
}


const rotationsLens = {
  get : (shapeBuffer                  ) => shapeBuffer.pendingRotations
, set : (shapeBuffer, pendingRotations) => shapeBuffer.pendingRotations = pendingRotations
}


const scalesLens = {
  get : (shapeBuffer               ) => shapeBuffer.pendingScales
, set : (shapeBuffer, pendingScales) => shapeBuffer.pendingScales = pendingScales
}


const colorsLens = {
  get : (shapeBuffer               ) => shapeBuffer.pendingColors
, set : (shapeBuffer, pendingColors) => shapeBuffer.pendingColors = pendingColors
}


function hasIdentifier(shapeBuffer, identifier) {
  return shapeBuffer.locationses.has(identifier)
}


function createShapeBuffer(gl, shapeProgram, primitiveMode, primitives) {
  return {
    shapeProgram     : shapeProgram
  , primitiveMode    : primitiveMode
  , mesh             : buildBuffer(gl, primitives, shapeProgram.meshDescription)
  , vertexCount      : primitives.length
  , instanceCount    : 0
  , positions        : null
  , rotations        : null
  , scales           : null
  , colors           : null
  , size             : 0
  , empties          : new Set()
  , locationses      : new Map()
  , pendingPositions : new Map()
  , pendingRotations : new Map()
  , pendingScales    : new Map()
  , pendingColors    : new Map()
  , pendingSize      : 0
  }
}


function destroyShapeBuffer(gl, shapeBuffer) {
  gl.deleteBuffer(shapeBuffer.mesh)
  if (shapeBuffer.size == 0)
    return
  [
    shapeBuffer.positions
  , shapeBuffer.rotations
  , shapeBuffer.scales
  , shapeBuffer.colors
  ].forEach(gl.deleteBuffer)
}


function insertPositions(identifier, positions, shapeBuffer) {
  positions.map((position) => insertPosition(shapeBuffer, identifier, position))
}


function insertPosition(shapeBuffer, identifier, vertex) {

  let empties1 = shapeBuffer.empties
  let pendingSize1 = shapeBuffer.pendingSize
  if (empties1.size == 0) {
    empties1 = new Set(Array.from({length: shapeBuffer.pendingSize % 2 + 1}, (v, k) => pendingSize1 + k))
    pendingSize1 = pendingSize1 + empties1.size
  }

  let location = null
  empties1.forEach(function(x) {location = location == null ? x : Math.min(location, x)})
  empties1.delete(location)

  console.debug("insertPosition: identifier =", identifier, ", vertex =", vertex)
  console.debug("insertPosition: location = ", location, ", pendingSize1 =", pendingSize1, ", empties1 =", empties1)
  shapeBuffer.empties = empties1
  shapeBuffer.pendingSize = pendingSize1
  shapeBuffer.pendingPositions.set(location, vertex)
  if (!shapeBuffer.locationses.has(identifier))
    shapeBuffer.locationses.set(identifier, new Set())
  shapeBuffer.locationses.get(identifier).add(location)

}


function updatePositions(identifier, positions, shapeBuffer) {
  console.debug("updatePositions: identifier =", identifier, ", positions =", positions)
  updateAttributes(positionsLens, identifier, positions, shapeBuffer)
}


function updateRotations(identifier, rotations, shapeBuffer) {
  console.debug("updateRotations: identifier =", identifier, ", rotations =", rotations)
  updateAttributes(rotationsLens, identifier, rotations, shapeBuffer)
}


function updateScales(identifier, scales, shapeBuffer) {
  console.debug("updateScales: identifier =", identifier, ", scales =", scales)
  updateAttributes(scalesLens, identifier, scales, shapeBuffer)
}


function updateAttributes(field, identifier, values, shapeBuffer) {
  if (shapeBuffer.locationses.has(identifier)) {
    const locations = shapeBuffer.locationses.get(identifier)
    const revisions = field.get(shapeBuffer)
    let i = 0
    locations.forEach((location) => revisions.set(location, values[i++]))
    field.set(shapeBuffer, revisions)
  }
}


function updateColor(identifier, color, shapeBuffer) {
  console.debug("updateColor: identifier =", identifier, ", color =", color)
  updateAttribute(colorsLens, identifier, color, shapeBuffer)
}


function updateAttribute(field, identifier, value, shapeBuffer) {
  if (shapeBuffer.locationses.has(identifier)) {
    const locations = shapeBuffer.locationses.get(identifier)
    const revisions = field.get(shapeBuffer)
    locations.forEach((location) => revisions.set(location, value))
    field.set(shapeBuffer, revisions)
  }
}


function deleteInstance(shapeBuffer, identifier) {
  if (shapeBuffer.locationses.has(identifier)) {
    const locations = shapeBuffer.locationses
    locations.forEach(function(location) {
      shapeBuffer.empties.add(location)
      shapeBuffer.pendingScales.set(location, zeroScale)
    })
    shapeBuffer.locationses.delete(identifier)
  }
}


function prepareShapeBuffer(gl, shapeBuffer) {
  console.debug("prepareShapeBuffer")
  expandShapeBuffer(gl, shapeBuffer)
  updateShapeBuffer(gl, shapeBuffer)
}


function updateShapeBuffer(gl, shapeBuffer) {

  if (shapeBuffer.pendingPositions.size +
      shapeBuffer.pendingRotations.size +
      shapeBuffer.pendingScales.size    +
      shapeBuffer.pendingColors     == 0)
    return

  console.debug("updateShapeBuffer")

  updateBuffer(gl, shapeBuffer.pendingPositions, shapeBuffer.positions, shapeBuffer.shapeProgram.positionsDescription)
  updateBuffer(gl, shapeBuffer.pendingRotations, shapeBuffer.rotations, shapeBuffer.shapeProgram.rotationsDescription)
  updateBuffer(gl, shapeBuffer.pendingScales   , shapeBuffer.scales   , shapeBuffer.shapeProgram.scalesDescription   )
  updateBuffer(gl, shapeBuffer.pendingColors   , shapeBuffer.colors   , shapeBuffer.shapeProgram.colorsDescription   )

  let location = null
  shapeBuffer.locationses.forEach((xs, _) =>
    xs.forEach(
      function(x) {
        location = Math.max(location, x)
      }
    )
  )

  shapeBuffer.instanceCount    = location + 1
  shapeBuffer.pendingPositions = new Map()
  shapeBuffer.pendingRotations = new Map()
  shapeBuffer.pendingScales    = new Map()
  shapeBuffer.pendingColors    = new Map()

}


function expandShapeBuffer(gl, shapeBuffer) {

  const size        = shapeBuffer.size
  const pendingSize = shapeBuffer.pendingSize

  if (pendingSize <= size)
    return

  function replicate(value) {
    return Array.from({length: pendingSize}, (v, k) => value)
  }

  console.debug("expandShapeBuffer: size =", size, ", pendingSize =", pendingSize)

  if (size == 0) {
    shapeBuffer.positions = buildBuffer(gl, replicate(zeroPosition), shapeBuffer.shapeProgram.positionsDescription)
    shapeBuffer.rotations = buildBuffer(gl, replicate(zeroRotation), shapeBuffer.shapeProgram.rotationsDescription)
    shapeBuffer.scales    = buildBuffer(gl, replicate(zeroScale   ), shapeBuffer.shapeProgram.scalesDescription   )
    shapeBuffer.colors    = buildBuffer(gl, replicate(zeroColor   ), shapeBuffer.shapeProgram.colorsDescription   )
  } else {
    shapeBuffer.positions = expandBuffer(gl, shapeBuffer.shapeProgram.positionsDescription, size, pendingSize, shapeBuffer.positions)
    shapeBuffer.rotations = expandBuffer(gl, shapeBuffer.shapeProgram.rotationsDescription, size, pendingSize, shapeBuffer.rotations)
    shapeBuffer.scales    = expandBuffer(gl, shapeBuffer.shapeProgram.scalesDescription   , size, pendingSize, shapeBuffer.scales   )
    shapeBuffer.colors    = expandBuffer(gl, shapeBuffer.shapeProgram.colorsDescription   , size, pendingSize, shapeBuffer.colors   )
  }

  shapeBuffer.size = pendingSize

}


function drawInstances(gl, shapeBuffer, projection, modelView) {

  if (shapeBuffer.size == 0)
    return

  console.debug("drawInstances: vertices =", shapeBuffer.vertexCount, ", instances =", shapeBuffer.instanceCount)

  const shapeProgram = shapeBuffer.shapeProgram

  Program.selectShapeProgram(gl, shapeProgram)

  Program.bindMesh     (gl, shapeProgram, shapeBuffer.mesh     )
  Program.bindPositions(gl, shapeProgram, shapeBuffer.positions)
  Program.bindRotations(gl, shapeProgram, shapeBuffer.rotations)
  Program.bindScales   (gl, shapeProgram, shapeBuffer.scales   )
  Program.bindColors   (gl, shapeProgram, shapeBuffer.colors   )

  Program.setProjectionModelView(gl, shapeProgram, projection, modelView)

  console.debug("drawInstances: gl.drawArraysInstanced")
  gl.drawArraysInstanced(shapeBuffer.primitiveMode, 0, shapeBuffer.vertexCount, shapeBuffer.instanceCount)

}


function buildBuffer(gl, primitives, description) {

  console.debug("buildBuffer: description = ", description, ", primitives ="  , primitives)

  const bufferObject = gl.createBuffer()

  gl.bindBuffer(gl.ARRAY_BUFFER, bufferObject)

  const components = description.components
  const bytes = description.isFloat                  ?
    new Float32Array(primitives.length * components) :
    new Uint32Array (primitives.length * components)
  if (components == 1)
    for (let i = 0; i < primitives.length; ++i)
      bytes[i] = primitives[i]
  else
    for (let i = 0; i < primitives.length; ++i)
      for (let j = 0; j < components; ++j)
        bytes[i * components + j] = primitives[i][j]
  gl.bufferData(gl.ARRAY_BUFFER, bytes, gl.DYNAMIC_DRAW)

  gl.bindBuffer(gl.ARRAY_BUFFER, null)

  return bufferObject
}


function updateBuffer(gl, updates, bufferObject, description) {

  console.debug("updateBuffer: description = ", description, ", updates ="     , updates)

  const components = description.components
  const bytes = description.isFloat ?
    new Float32Array(components)    :
    new Uint32Array (components)

  gl.bindBuffer(gl.ARRAY_BUFFER, bufferObject)

  updates.forEach(function(value, location) {
    if (components == 1)
      bytes[0] = value
    else
      for (let j = 0; j < components; ++j)
        bytes[j] = value[j]
    gl.bufferSubData(gl.ARRAY_BUFFER, components * location, bytes)
  })

  gl.bindBuffer(gl.ARRAY_BUFFER, null)

}


function expandBuffer(gl, description, oldSize, newSize, oldBufferObject) {

  console.debug("expandBuffer: description = ", description, ", oldSize =", oldSize, ", newSize =", newSize)

  const bytes = 4 // 32-bit elements.
  const components = description.components

  const newBufferObject = gl.createBuffer()
  gl.bindBuffer(gl.ARRAY_BUFFER, newBufferObject)
  gl.bufferData(newBufferObject, bytes * components * newSize, gl.DYNAMIC_DRAW)

  gl.bindBuffer(gl.COPY_READ_BUFFER, oldBufferObject)
  gl.copyBufferSubData(gl.COPY_READY_BUFFER, gl.ARRAY_BUFFER, 0, 0, bytes * components * oldSize)
  gl.bindBuffer(gl.ARRAY_BUFFER, null)

  gl.bindBuffer(gl.COPY_READY_BUFFER, null)
  gl.deleteBuffer(oldBufferObject)

  return newBufferObject

}


module.exports = {
  createShapeBuffer  : createShapeBuffer
, prepareShapeBuffer : prepareShapeBuffer
, destroyShapeBuffer : destroyShapeBuffer
, hasIdentifier      : hasIdentifier
, insertPositions    : insertPositions
, updatePositions    : updatePositions
, updateRotations    : updateRotations
, updateScales       : updateScales
, updateColor        : updateColor
, deleteInstance     : deleteInstance
, drawInstances      : drawInstances

, buildBuffer        : buildBuffer
}
